### PR TITLE
rocksdb: add compression type zstd and lz4

### PR DESCRIPTION
**Pull request by neverchanje**
In brief, if have lots of free CPU, **ZSTD** is a good chooice; otherwise, **LZ4** is recommended.

Compressor name | Ratio | Compression(MB/s) | Decompress(MB/s)
-- | -- | -- | --
zstd 1.3.4 -1 | 2.877 | 470 | 1380
zlib 1.2.11 -1 | 2.743 | 110 | 400
brotli 1.0.2 -0 | 2.701 | 410 | 430
quicklz 1.5.0 -1 | 2.238 | 550 | 710
lzo1x 2.09 -1 | 2.108 | 650 | 830
lz4 1.8.1 | 2.101 | 750 | 3700
snappy 1.1.4 | 2.091 | 530 | 1800
lzf 3.6 -1 | 2.077 | 400 | 860

RocksDB gives some suggestion [here](https://github.com/facebook/rocksdb/wiki/Compression):
>Use options.compression to specify the compression to use. By default it is Snappy. We believe LZ4 is almost always better than Snappy. We leave Snappy as default to avoid unexpected compatibility problems to previous users. LZ4/Snappy is lightweight compression so it usually strikes a good balance between space and CPU usage.

>If you want to further reduce the in-memory and have some free CPU to use, you can try to set a heavy-weight compression in the latter by setting options.bottommost_compression. The bottommost level will be compressed using this compression style. Usually the bottommost level contains majority of the data, so users get almost optimal space setting, without paying CPU for compress all the data ever flowing to any level. We recommend ZSTD. If it is not available, Zlib is the second choice.

>If you want have a lot of free CPU and want to reduce not just space but write amplification too, try to set options.compression to heavy weight compression type. We recommend ZSTD. Use Zlib if it is not available.

as we can see from the above table, lz4 is a compression algorithm with more moderate ratio and speed than zstd, so we also take it into consideration.

In this PR we disabled rocksdb's link with bzip2 (because we never use this lib for compression before), and add links with zstd and lz4. This two libs will be packed up using `./run.sh pack_server` and `./run.sh pack_tools`.

### CODE REVIEW

**Comment by qinzuoyan** [comment]
lz4 is a compression algorithm with more moderate ratio?

**Comment by acelyc111** [comment]
https://github.com/facebook/rocksdb/wiki/Compression
”LZ4/Snappy is lightweight compression so it usually strikes a good balance between space and CPU usage.“
ZSTD压缩率最高，但也更耗CPU。

**Comment by qinzuoyan** [comment]
看起来lz4无论在压缩率还是速度上都优于snappy，为何感觉snappy更受欢迎呢？

**Comment by neverchanje** [comment]
> Use options.compression to specify the compression to use. By default it is Snappy. We believe LZ4 is almost always better than Snappy. We leave Snappy as default to avoid unexpected compatibility problems to previous users. LZ4/Snappy is lightweight compression so it usually strikes a good balance between space and CPU usage.

是为了数据兼容性。snappy 跟 zlib 比起来确实是更好一些，但是 lz4/zstd 比 snappy 更年轻。
lz4 官方 benchmark 结果：

Compressor | Ratio | Compression | Decompression
-- | -- | -- | --
memcpy | 1.000 | 13100 MB/s | 13100 MB/s
LZ4 default (v1.8.2) | 2.101 | 730 MB/s | 3900 MB/s
LZO 2.09 | 2.108 | 630 MB/s | 800 MB/s
QuickLZ 1.5.0 | 2.238 | 530 MB/s | 720 MB/s
Snappy 1.1.4 | 2.091 | 525 MB/s | 1750 MB/s
Zstandard 1.3.4 -1 | 2.877 | 470 MB/s | 1380 MB/s
LZF v3.6 | 2.073 | 380 MB/s | 840 MB/s
zlib deflate 1.2.11 -1 | 2.730 | 100 MB/s | 380 MB/s
LZ4 HC -9 (v1.8.2) | 2.721 | 40 MB/s | 3920 MB/s
zlib deflate 1.2.11 -6 | 3.099 | 34 MB/s | 410 MB/s

**Comment by qinzuoyan** [comment]
看来以后我们线上默认用lz4更合适

File: src/server/pegasus_server_impl.cpp [hunk 1]
```diff
@@ -181,11 +181,15 @@ pegasus_server_impl::pegasus_server_impl(dsn::replication::replica *r)
         "pegasus.server",
         "rocksdb_compression_type",
         "snappy",
-        "rocksdb options.compression, default snappy. Supported: none, snappy.");
+        "rocksdb options.compression, default snappy. Supported: none, snappy, zstd, lz4.");
     if (compression_str == "none") {
         _db_opts.compression = rocksdb::kNoCompression;
     } else if (compression_str == "snappy") {
         _db_opts.compression = rocksdb::kSnappyCompression;
+    } else if (compression_str == "zstd") {
```
Comment (qinzuoyan) [inline src/server/pegasus_server_impl.cpp]: 这里仅仅增加这几行代码还不够吧，还要在依赖库的链接上把这两种压缩也加进来

**Comment by acelyc111** [comment]
> 看来以后我们线上默认用lz4更合适

之前看了下线上机器，很多都没装lz4库。要上的话需要提前都装好

**Comment by qinzuoyan** [comment]
或者在pack_server的时候，把lz4/zstd库都带上

File: scripts/pack_server.sh [hunk 2]
```diff
@@ -102,7 +102,8 @@ copy_file `get_system_lib server snappy` ${pack}/bin
 copy_file `get_system_lib server crypto` ${pack}/bin
 copy_file `get_system_lib server ssl` ${pack}/bin
 copy_file `get_system_lib server aio` ${pack}/bin
-copy_file `get_system_lib server bz2` ${pack}/bin
+copy_file `get_system_lib server zstd` ${pack}/bin
```
Comment (qinzuoyan) [inline scripts/pack_server.sh]: 最好在服务器机器都测试下（譬如升级c4tst-benchmark集群）：看是否能找到库，正常启动。
另外通过pack_tools得到的shell工具也要在服务器机器环境上测试下，看是否需要依赖这两个库；如果需要，也要pack进来。

File: .travis.yml [hunk 3]
```diff
@@ -19,9 +26,9 @@ before_install:
     - cd ..
 
 install:
-    - sudo apt-get -f install
-    - sudo apt-get -y install libsnappy-dev
-    - sudo apt-get -y install libgflags-dev
+    - wget https://github.com/facebook/zstd/archive/v1.3.7.zip; unzip v1.3.7; cd zstd-1.3.7;
+    - mkdir cmake-build; cd cmake-build; cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_INSTALL_LIBDIR=lib -DZSTD_BUILD_PROGRAMS=OFF ../build/cmake; sudo make install -j8;
```
Comment (qinzuoyan) [inline .travis.yml]: 为什么多个shell语句有放在一行？分开放更合理吧？

**Pull request merged**